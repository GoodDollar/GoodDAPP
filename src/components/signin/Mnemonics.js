// @flow
//eslint-disable-next-line

import React, { useEffect, useRef, useState } from 'react'
import { get } from 'lodash'
import bip39 from 'bip39-light'
import AsyncStorage from '../../lib/utils/asyncStorage'
import { IS_LOGGED_IN } from '../../lib/constants/localStorage'
import logger from '../../lib/logger/pino-logger'
import { ExceptionCategory } from '../../lib/logger/exceptions'
import { withStyles } from '../../lib/styles'
import { useDialog, useErrorDialog } from '../../lib/undux/utils/dialog'
import { getFirstWord } from '../../lib/utils/getFirstWord'
import { userExists } from '../../lib/login/userExists'
import Text from '../common/view/Text'
import Section from '../common/layout/Section'
import { showSupportDialog } from '../common/dialogs/showSupportDialog'
import SuccessAnimation from '../common/animations/Success'
import CustomButton from '../common/buttons/CustomButton'
import InputText from '../common/form/InputText'
import { CLICK_BTN_RECOVER_WALLET, fireEvent, RECOVER_FAILED, RECOVER_SUCCESS } from '../../lib/analytics/analytics'
import Wrapper from '../common/layout/Wrapper'

const TITLE = 'Recover'
const log = logger.child({ from: TITLE })
const MAX_WORDS = 12

const Mnemonics = ({ screenProps, navigation, styles }) => {
  //lazy load heavy wallet stuff for fast initial app load (part of initial routes)
  const mnemonicsHelpers = import('../../lib/wallet/SoftwareWalletProvider')
  const [mnemonics, setMnemonics] = useState()
  const [isRecovering, setRecovering] = useState(false)
  const [isSubmitBlocked, setSubmitBlocked] = useState(true)
  const [showDialog] = useDialog()
  const [errorMessage, setErrorMessage] = useState()
  const [showErrorDialog, hideDialog] = useErrorDialog()
  const input = useRef()

  AsyncStorage.removeItem('GD_web3Token')

  const handleChange = (mnemonics: string) => {
    log.info({ mnemonics })
    const splitted = mnemonics.split(' ').filter(o => o)
    if (splitted.length > MAX_WORDS) {
      setErrorMessage('Your pass phrase appears to be incorrect.')
    } else {
      setErrorMessage(null)
    }
    if (splitted.length === MAX_WORDS) {
      setSubmitBlocked(false)
    } else {
      setSubmitBlocked(true)
    }
    setMnemonics(mnemonics)
  }

  const recover = async () => {
    //required to wallet and storage are reinitialized
    const curVersion = await AsyncStorage.getItem('GD_version')
    await AsyncStorage.clear()
    AsyncStorage.setItem('GD_version', curVersion)
    input.current.blur()
    setRecovering(true)
    fireEvent(CLICK_BTN_RECOVER_WALLET)

    const showError = () => {
      const error = new Error('Incorrect pass phrase received')

      log.error('Wallet recover failed', error.message, error, {
        mnemonics,
        category: ExceptionCategory.Human,
        dialogShown: true,
      })
      showErrorDialog('Your pass phrase appears\nto be incorrect.', undefined, {
        boldMessage: 'Please check it and try again.',
      })
    }

    if (!mnemonics || !bip39.validateMnemonic(mnemonics)) {
      fireEvent(RECOVER_FAILED, { invalidMnemonics: true })
      setRecovering(false)
      showError()
      return
    }

    const { getMnemonics, saveMnemonics } = await mnemonicsHelpers
    const prevMnemonics = await getMnemonics()

    try {
      // We need to try to get a new address using new mnenonics
      //eslint-disable-next-line
      await saveMnemonics(mnemonics)

      // We validate that a user was registered for the specified mnemonics
      const { exists, fullName } = await userExists(mnemonics)
      log.debug('userExists result:', { exists, fullName })

      if (exists) {
        await AsyncStorage.setItem(IS_LOGGED_IN, true)
        const incomingRedirectUrl = get(navigation, 'state.params.redirect', '/')
        const firstName = getFirstWord(fullName)
        showDialog({
          visible: true,
          image: <SuccessAnimation />,
          buttons: [{ text: 'Yay!' }],
          message: `Hi ${firstName},\nyour wallet was recovered successfully`,
          onDismiss: () => (window.location = incomingRedirectUrl),
        })
        fireEvent(RECOVER_SUCCESS)

        // There is no error and Profile exists. Reload screen to start with users mnemonics
        // window.location = incomingRedirectUrl
      } else {
        fireEvent(RECOVER_FAILED, { noProfileFound: true })
        await saveMnemonics(prevMnemonics)
        showError()
      }
    } catch (e) {
      fireEvent(RECOVER_FAILED, { unexpected: true })
      log.error('recover mnemonics failed', e.message, e, { dialogShown: true })
      saveMnemonics(prevMnemonics)
      showSupportDialog(showErrorDialog, hideDialog, screenProps.push)
    } finally {
      setRecovering(false)
    }
  }
  const handleEnter = (event: { nativeEvent: { key: string } }) => {
    if (event.nativeEvent.key === 'Enter') {
      recover()
    }
  }

  const incomingMnemonic = get(navigation, 'state.params.mnemonic', undefined)

  useEffect(() => {
    if (incomingMnemonic) {
      handleChange(incomingMnemonic)
    }
  }, [])

  const web3HasWallet = get(navigation, 'state.params.web3HasWallet')

  return (
    <Wrapper style={styles.mainWrapper}>
      <Section grow={5} style={styles.wrapper}>
        <Section.Stack grow style={styles.instructions} justifyContent="space-around">
          <Text fontWeight="medium" fontSize={22}>
            {'Please enter your\n12-word pass phrase:'}
          </Text>
          {web3HasWallet && (
            <Text color="gray80Percent" fontSize={14}>
              Looks like you already have a wallet. Please recover it to continue
            </Text>
          )}
        </Section.Stack>
        <Section.Stack grow={4} justifyContent="space-between">
          <Section.Row justifyContent="center">
            <InputText
              value={mnemonics}
              onChangeText={handleChange}
              error={errorMessage}
              onKeyPress={handleEnter}
              getRef={input}
              showCleanAdornment
              autoFocus
            />
          </Section.Row>
        </Section.Stack>
        <Section.Row style={styles.instructions} justifyContent="space-around">
          <Text color="gray80Percent" fontSize={14}>
            {'You can copy-paste all of it at once\n from your '}
            <Text color="gray80Percent" fontSize={14} fontWeight="bold">
              {'backup email'}
            </Text>
          </Text>
        </Section.Row>
        <Section.Stack grow style={styles.bottomContainer} justifyContent="flex-end">
          <CustomButton style={styles.buttonLayout} onPress={recover} disabled={isSubmitBlocked || isRecovering}>
            Recover my wallet
          </CustomButton>
        </Section.Stack>
      </Section>
    </Wrapper>
  )
}

Mnemonics.navigationOptions = {
  title: TITLE,
}

const mnemonicsStyles = ({ theme }) => ({
  wrapper: {
    borderRadius: 0,
  },
  instructions: {
    marginVertical: theme.paddings.defaultMargin,
  },
  buttonLayout: {
    marginTop: 20,
    marginBottom: 20,
  },
  bottomContainer: {
    maxHeight: 80,
    minHeight: 80,
  },
  mainWrapper: {
    backgroundImage: 'none',
    backgroundColor: 'none',
  },
})

export default withStyles(mnemonicsStyles)(Mnemonics)
