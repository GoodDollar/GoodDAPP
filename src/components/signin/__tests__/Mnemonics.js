import React from 'react'
import renderer from 'react-test-renderer'
import Mnemonics from '../Mnemonics'
import { withThemeProvider } from '../../../__tests__/__util__'

const MnemonicsTheme = withThemeProvider(Mnemonics)

describe('SignIn - Mnemonics', () => {
  it('matches snapshot', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<MnemonicsTheme />)))
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
