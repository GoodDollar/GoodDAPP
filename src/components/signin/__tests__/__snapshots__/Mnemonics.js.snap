// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn - Mnemonics matches snapshot 1`] = `
<div
  className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
>
  <div
    className="css-view-1dbjc4n r-backgroundColor-1b3fm86 r-display-6koalj r-flexDirection-eqz5dr r-flexGrow-16y2uox r-maxHeight-87ditf r-overflowX-lltvgl r-overflowY-1rnoaur r-padding-edyy15 r-position-bnwqim r-width-13qz1uu"
    data-name="viewWrapper"
    style={
      Object {
        "backgroundImage": "linear-gradient(3.141592653589793rad,#00AFFF,#2DC0F7,#28C0EF,#23C0E7,#1EC1DF,#19C1D7,#14C1CF,#0FC2C7,#0FC2C7,#0AC2BF,#05C2B7,#00C3AF)",
      }
    }
  >
    <div
      className="css-view-1dbjc4n r-backgroundColor-14lw9ot r-borderRadius-17gur6a r-paddingHorizontal-1e081e0 r-paddingVertical-1yzf0co"
      style={
        Object {
          "WebkitBoxFlex": 5,
          "WebkitFlexGrow": 5,
          "flexGrow": 5,
          "msFlexPositive": 5,
        }
      }
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-flexGrow-16y2uox r-marginVertical-c8eef1"
        style={
          Object {
            "WebkitBoxPack": "justify",
            "WebkitJustifyContent": "space-around",
            "justifyContent": "space-around",
            "msFlexPack": "distribute",
          }
        }
      >
        <div
          className="css-text-901oao r-color-cddv6f r-fontFamily-vnw8o6 r-fontSize-hiw28u r-fontWeight-majxgm r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
          dir="auto"
        >
          Please enter your
          

          12-word pass phrase:
        </div>
      </div>
      <div
        className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
        style={
          Object {
            "WebkitBoxFlex": 4,
            "WebkitFlexGrow": 4,
            "flexGrow": 4,
            "msFlexPositive": 4,
          }
        }
      >
        <div
          className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1777fci"
        >
          <div
            className="css-view-1dbjc4n r-marginBottom-5oul0u r-width-13qz1uu"
          >
            <div
              className="css-view-1dbjc4n r-marginBottom-5oul0u r-width-13qz1uu"
            >
              <input
                autoCapitalize="sentences"
                autoComplete="on"
                autoCorrect="on"
                autoFocus={true}
                className="css-textinput-11aywtz r-backgroundColor-14lw9ot r-borderBottomColor-1htqvjr r-borderBottomWidth-qklmqi r-color-13sdq65 r-fontFamily-6jnb14 r-fontSize-hiw28u r-fontWeight-1it3c9n r-paddingHorizontal-1dye5f7 r-paddingVertical-oyd9sg r-placeholderTextColor-1xktb6b r-textAlign-q4m81j"
                dir="auto"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onSelect={[Function]}
                onTouchStart={[Function]}
                readOnly={false}
                spellCheck={true}
                type="text"
              />
              <div
                className="css-view-1dbjc4n r-bottom-1ebgqk7 r-cursor-1loqt21 r-paddingTop-tskmnb r-position-u8s1d r-right-3mc0re r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-zIndex-184en5c"
                onClick={[Function]}
                onContextMenu={[Function]}
                onKeyDown={[Function]}
                style={
                  Object {
                    "WebkitTransitionDuration": "0s",
                    "transitionDuration": "0s",
                  }
                }
                tabIndex="0"
              >
                <div
                  className="css-text-901oao r-color-13sdq65 r-fontWeight-1it3c9n"
                  dir="auto"
                  style={
                    Object {
                      "fontFamily": "gooddollar",
                      "fontSize": "16px",
                      "fontStyle": "normal",
                    }
                  }
                >
                  î Ž
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-marginVertical-c8eef1"
        style={
          Object {
            "WebkitBoxPack": "justify",
            "WebkitJustifyContent": "space-around",
            "justifyContent": "space-around",
            "msFlexPack": "distribute",
          }
        }
      >
        <div
          className="css-text-901oao r-color-vaa0vc r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
          dir="auto"
        >
          You can copy-paste all of it at once
 from your 
          <span
            className="css-text-901oao css-textHasAncestor-16my406 r-color-vaa0vc r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-b88u0q r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
          >
            backup email
          </span>
        </div>
      </div>
      <div
        className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-flexGrow-16y2uox r-maxHeight-14c2kc1 r-minHeight-mjyw6"
        style={
          Object {
            "WebkitBoxPack": "end",
            "WebkitJustifyContent": "flex-end",
            "justifyContent": "flex-end",
            "msFlexPack": "end",
          }
        }
      >
        <div
          className="css-view-1dbjc4n r-alignItems-1oszu61 r-display-6koalj r-justifyContent-1777fci"
          style={
            Object {
              "backgroundColor": "rgba(23,53,102,0.40)",
              "borderBottomColor": "rgba(0,175,255,1.00)",
              "borderBottomLeftRadius": "50px",
              "borderBottomRightRadius": "50px",
              "borderBottomStyle": "solid",
              "borderBottomWidth": "0px",
              "borderLeftColor": "rgba(0,175,255,1.00)",
              "borderLeftStyle": "solid",
              "borderLeftWidth": "0px",
              "borderRightColor": "rgba(0,175,255,1.00)",
              "borderRightStyle": "solid",
              "borderRightWidth": "0px",
              "borderTopColor": "rgba(0,175,255,1.00)",
              "borderTopLeftRadius": "50px",
              "borderTopRightRadius": "50px",
              "borderTopStyle": "solid",
              "borderTopWidth": "0px",
              "marginBottom": "20px",
              "marginLeft": "0px",
              "marginRight": "0px",
              "marginTop": "20px",
              "minHeight": "44px",
              "minWidth": "64px",
              "paddingBottom": "0px",
              "paddingLeft": "0px",
              "paddingRight": "0px",
              "paddingTop": "0px",
            }
          }
        >
          <div
            aria-disabled={true}
            className="css-view-1dbjc4n r-flex-13awgt0 r-overflow-1udh08x r-position-bnwqim"
            onClick={[Function]}
            onContextMenu={[Function]}
            onKeyDown={[Function]}
            role="button"
            style={
              Object {
                "WebkitAlignSelf": "stretch",
                "alignSelf": "stretch",
                "borderBottomLeftRadius": "50px",
                "borderBottomRightRadius": "50px",
                "borderTopLeftRadius": "50px",
                "borderTopRightRadius": "50px",
                "msFlexItemAlign": "stretch",
                "msGridRowAlign": "stretch",
              }
            }
            tabIndex="-1"
          >
            <div
              className="css-view-1dbjc4n r-alignItems-1awozwy r-flex-13awgt0 r-flexDirection-18u37iz r-justifyContent-1777fci"
              style={
                Object {
                  "WebkitAlignSelf": "stretch",
                  "alignSelf": "stretch",
                  "msFlexItemAlign": "stretch",
                  "msGridRowAlign": "stretch",
                }
              }
            >
              <div
                className="css-view-1dbjc4n r-marginHorizontal-1ydw1k6 r-marginVertical-712mwr"
              >
                <div
                  className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-flexDirection-18u37iz"
                >
                  <div
                    className="css-text-901oao r-alignItems-1awozwy r-color-jwli3a r-display-6koalj r-fontFamily-vnw8o6 r-fontSize-ubezar r-fontWeight-majxgm r-justifyContent-1777fci r-letterSpacing-oxtfae r-lineHeight-qvk6io r-minHeight-1ceczpf r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-tsynxw"
                    dir="auto"
                  >
                    Recover my wallet
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
